Fix hard-coded byte alignment in linker scripts

peter-mitsis opened the pull request:
Most instances of the ITERABLE_SECTION_xxx() macros used a hard-coded value of 4 for the sub-alignment. As that has the potential to cause issues with 64-bit platforms that should use 8 instead, those cases now use the new macro Z_LINK_ITERABLE_SUBALIGN instead as it is defined as 4 for 32-bit platforms and 8 for 64-bit platforms.

Fixes #69859

peter-mitsis committed:
linker: Add Z_LINK_ITERABLE_SUBALIGN

The macro Z_LINK_ITERABLE_SUBALIGN is used by linker scripts to
select the default subalignment based on whether it is 32/64-bit
architecture. 64-bit architectures set the default subalignment
to 8, while 32-bit architectures set it to 4.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
linker: tests: Use Z_LINK_ITERABLE_SUBALIGN

Updates the linker script fragments under 'tests' to use
Z_LINK_ITERABLE_SUBALIGN for the subalignment instead of
a hardcoded value of 4.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
linker: samples: Use Z_LINK_ITERABLE_SUBALIGN

Updates the linker script fragments under 'samples' to use
Z_LINK_ITERABLE_SUBALIGN for the subalignment instead of
a hardcoded value of 4.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
linker: subsys: Use Z_LINK_ITERABLE_SUBALIGN

Updates the linker script fragments under 'subsys' to use
Z_LINK_ITERABLE_SUBALIGN for the subalignment instead of
a hardcoded value of 4.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
linker: drivers: Use Z_LINK_ITERABLE_SUBALIGN

Updates the linker script fragments under 'drivers' to use
Z_LINK_ITERABLE_SUBALIGN for the subalignment instead of
a hardcoded value of 4.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

pdgendt commented on include/zephyr/linker/iterable_sections.h:
```diff
@@ -32,6 +32,12 @@
 	*(SORT_BY_NAME(._##struct_type.static.*)); \
 	_CONCAT(_##struct_type, _list_end) = .
 
+#ifdef CONFIG_64BIT
+#define Z_LINK_ITERABLE_SUBALIGN  8
+#else
+#define Z_LINK_ITERABLE_SUBALIGN  4
+#endif
```
Alternatively a `kconfig` symbol could be created directly

```
config ITERABLE_SECTION_SUBALIGN
	int
	default 8 if 64BIT
	default 4
```

pdgendt replied:
See #72373

glarsennordic commented:
`conn_mgr` changes LGTM as long as tests pass. I am neutral between the KConfig and the C Macro approach.

Part of me wonders if we should look into removing the argument, or providing versions of the API that don't need the argument, but that _**would**_ be a breaking API change, so probably better not to do that for now. (Not to mention the singular iterable section in Zephyr (that I cold find) that uses an align other than 4, `zbus_channel_observation_mask`)

```c
ITERABLE_SECTION_RAM(zbus_channel_observation_mask, 1)
```

glarsennordic commented:
Still LGTM, not approving so that others get a chance to review

pdgendt commented:
> Still LGTM, not approving so that others get a chance to review

I think the CI failures should be looked at and get fixed too.

peter-mitsis committed:
tests: Update iterable sections test

Updates the test_ram/test_rom structures to use a 'long' instead
of an 'int' so that the structure sizes will be 4-bytes on 32-bit
platforms and 8-bytes on 64-bit platforms.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

pdgendt commented:
> Adding the DNM tag to ensure that no one merges this (yet). There is something weird going on with many of the logging tests on 64-bit platforms with this change set. I am trying to track down the cause.

Not sure if this is for the logging tests only, `tests/misc/iterable_sections` also fails on `native_sim/native/64` but not on `native_sim`.
My guess would be that there is some mismatch in where padding is added in the linker vs compiler. I had similar issues when trying to put `__packed` structs in linker sections.

pdgendt commented:
Upon closer inspection, the issue is actually in the `TYPE_SECTION_FOREACH` iterator, as it doesn't know about the struct alignment.

pdgendt commented:
Maybe we can add the alignment at the end of the struct area:

```c
#ifdef CONFIG_64BIT
#define Z_LINK_ITERABLE_SUBALIGN  8
#define Z_WORD(x) QUAD(x)
#else
#define Z_LINK_ITERABLE_SUBALIGN  4
#define Z_WORD(x) LONG(x)
#endif
// ...

#define ITERABLE_SECTION_ROM(struct_type, subalign) \
	SECTION_PROLOGUE(struct_type##_area,,SUBALIGN(subalign)) \
	{ \
		Z_LINK_ITERABLE(struct_type); \
		Z_WORD(subalign) \
	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
```

Which then can be accessed using the `_list_end` address, and compute the iterator increments based on the size and alignment?

peter-mitsis committed:
logging: Update log_source_dynamic_data struct

Updates the log_source_dynamic_data structure to be a multiple of 8
bytes in size on 64-bit platforms.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
There was a discrepancy between linker and compiler alignment for two of the structures in the previously observed failing CI tests. On 64-bit platforms, the linker was using 8-byte alignment, but the compiler was using 4-byte alignment. Consequently, when the TYPE_SECTION_FOREACH() was used, its iterator pointer was not pointing at the correct address when it was advanced to the next element.

This updated patch set should fix that by forcing the size of those two structures to be mulitples of 8 bytes, thereby disallowing any padding between successive instances of the structures within the relevant iterable sections.
